### PR TITLE
escrow: simplify, remove atomic swap functionality

### LINKED ISSUE #601: Simplify escrow
Points from https://github.com/iov-one/weave/issues/540

- remove hashlock package and middleware
- use Address instead of Condition for arbiter
- moveCoins helper from escrow and aswap can be moved to cash for reusability

Anything else?

### PR DESCRIPTION
**Pull request by ruseinov**
Fixes #601 
Currently this breaks binary compatibility by moving multisig message in place of deprecated preimage.

### CODE REVIEW

**Comment by ruseinov** [comment]
@ethanfrey I would assume that we should actually use `escr.Arbiter` instead of this now https://github.com/iov-one/weave/blob/1d14abe65c00ebede3cdd6911d9f5b84964a7d60/x/escrow/init.go#L50, correct?

**Comment by ethanfrey** [comment]
No. `escAddr` is the Address controlled by the escrow. Which is where the tokens should be stored (by the cash module)

File: cmd/bcpd/app/codec.proto [hunk 1]
```diff
@@ -23,10 +23,8 @@ import "github.com/iov-one/weave/x/validators/codec.proto";
 message Tx {
   cash.FeeInfo fees = 1;
   repeated sigs.StdSignature signatures = 2;
-  // Preimage for hashlock.
-  bytes preimage = 3;
   // ID of a multisig contract.
-  repeated bytes multisig = 4;
```
Comment (ethanfrey) [inline cmd/bcpd/app/codec.proto]: No need to break this. The idea was to keep all numbers constant.

I would just comment out 3 and write a comment `// 3 removed: previously hashock`. Please keep multisig at 4

File: cmd/bnsd/app/codec.proto [hunk 2]
```diff
@@ -27,10 +27,8 @@ import "github.com/iov-one/weave/x/validators/codec.proto";
 message Tx {
   cash.FeeInfo fees = 1;
   repeated sigs.StdSignature signatures = 2;
-  // Preimage for hashlock.
-  bytes preimage = 3;
   // ID of a multisig contract.
-  repeated bytes multisig = 4;
```
Comment (ethanfrey) [inline cmd/bnsd/app/codec.proto]: Same here. Please don't change field numbers is at all possible

**Comment by ruseinov** [comment]
> No. `escAddr` is the Address controlled by the escrow. Which is where the tokens should be stored (by the cash module)

I had a suspicion it would be so, good to know!

File: cmd/bcpd/app/codec.proto (hunk 1 above)
Comment (ruseinov) [reply cmd/bcpd/app/codec.proto]: I've added a general comment about gaps in field sequences as I don't think having a comment per deprecated field is a good maintenance strategy, no reason to pollute the code with deprecated stuff I suppose. Let me know if it's good enough.

**Comment by ruseinov** [comment]
> > No. `escAddr` is the Address controlled by the escrow. Which is where the tokens should be stored (by the cash module)
> 
> I had a suspicion it would be so, good to know!

By the way this is one point that I feel needs to be documented as it's kind of clear, but not really.

I will make sure escrow gets some helpful comments along the way.

**Comment by ruseinov** [comment]
@husio I have an issue with an acceptance test after changing escrow initialiser, if you have an idea for quick fix - please let me know:
```
--- FAIL: TestIssueNfts (0.01s)
panic: cannot decode hex: encoding/hex: invalid byte: U+006D 'm' [recovered]
	panic: cannot decode hex: encoding/hex: invalid byte: U+006D 'm'
--- FAIL: TestRevenueDistribution (0.07s)
panic: cannot decode hex: encoding/hex: invalid byte: U+006D 'm' [recovered]
	panic: cannot decode hex: encoding/hex: invalid byte: U+006D 'm'
```

File: cmd/bnsd/app/testdata/fixtures/app.go [hunk 3]
```diff
@@ -137,7 +137,7 @@ func appStateGenesis(keyAddress weave.Address) []byte {
 		"escrow": []interface{}{
 			dict{
 				"sender":    "0000000000000000000000000000000000000000",
-				"arbiter":   "multisig/usage/0000000000000001",
+				"arbiter":   weave.Condition("cond:multisig/usage/0000000000000001").Address(),
```
Comment (husio) [inline cmd/bnsd/app/testdata/fixtures/app.go]: [`cond` prefix is used by JSON deserializer](https://github.com/iov-one/weave/blob/master/conditions.go#L170-L179). When building a condition manually you should not prefix it. You can use [`NewCondition`](https://github.com/iov-one/weave/blob/master/conditions.go#L33) to build one.

Comment (ruseinov) [reply cmd/bnsd/app/testdata/fixtures/app.go]: Yeah, I know. It was a "clever" hack to make it pass, I'll fix that up properly now.

File: x/escrow/init_test.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ func TestGenesisKey(t *testing.T) {
           "whole": 123456789
         }
       ],
-      "arbiter": "foo/bar/636f6e646974696f6e64617461",
+      "arbiter": "0000000000000000000000000000000000000001",
```
Comment (ethanfrey) [inline x/escrow/init_test.go]: This is an odd address.
WHy not a more realistic one, like recipient

Comment (ethanfrey) [reply x/escrow/init_test.go]: Or better yet, keep the old condition and make it an address:

`cond:foo/bar/636f6e646974696f6e64617461`

I think we need to document our features much better if our own devs don't use them

**Pull request merged**